### ISSUE TITLE
Scoping rules are applied differently when a property declaration contains an assignment

### ISSUE THREAD

**Issue by RobertCraigie** [microsoft/pyright#5630]
**Describe the bug**

Pyright doesn't report any errors for a property declaration like this:
```python
balance: Optional[balance.Balance]
```

But will report errors for a property declaration like this:

```python
balance: Optional[balance.Balance] = None
```

**Code or Screenshots**

```python
# balance.py
class Balance:
  ...

# main.py
import balance
from typing import Optional

class PaymentCreateResponse:
    balance: Optional[balance.Balance] = None
```
```
main.py
  main.py:4:23 - error: Type of "balance" could not be determined because it refers to itself (reportGeneralTypeIssues)
  main.py:4:23 - error: Variable not allowed in type expression (reportGeneralTypeIssues)
  main.py:4:5 - error: Type of "balance" is partially unknown
    Type of "balance" is "Unknown | None" (reportUnknownVariableType)
  main.py:4:23 - error: Type of "Balance" is unknown (reportUnknownMemberType)
```

Note: I do appreciate that this code is fragile & recommended against, #3924, but it does work at runtime.

**VS Code extension or command-line**

Command-line: `v1.1.319`

I bisected this error & it started occurring in `v1.1.281`.

**Comment by erictraut** [comment]
Mypy also generates an error in this case.

I don't have that much interest in fixing this because, as you point out, it's really unsound. Even if it doesn't generate a runtime error, it should probably be considered a type checking error. It's definitely a type checking error if you use `from __future__ import annotations`, and that behavior will become the default in Python 3.13. So if someone is relying on such fragile behavior, I could argue that it's good to tell them about it now so they can fix their code.

I'll spend a little time investigating why the behavior is inconsistent between global and class scopes. If there's a really easy and straightforward fix, I might implement it. Otherwise I'll close this as "won't fix".

**Comment by RobertCraigie** [comment]
Ah it does actually error at runtime, looks like I was testing it incorrectly. Sorry for the invalid report & thanks for investigating!
